Feature/add logic model

bhaalgore opened the pull request:
Добавил модель, с помощью которой можно взаимодействовать с qml объектами.

Данный пр относится к данной задаче в трелло: https://trello.com/c/dE2l57j9

Примечание. Некорректные моменты в CMake листах исправлю при решении конфликтов, когда буду мержить в мастер то что написал Артем

bhaalgore committed:
Merge branch 'master' into feature/add-logic-model

bhaalgore committed:
separate notify_model from qml files

bhaalgore committed:
fix

Reensef commented on app/src/main.cpp:
```diff
@@ -1,11 +1,24 @@
-int main()
+#include <QGuiApplication>
+#include <QQmlApplicationEngine>
+#include "notify_model.h"
```
.h заголовки стоит писать рядом вместе

Reensef commented on app/src/main.cpp:
```diff
@@ -1,11 +1,24 @@
+#include "notify_model.h"
+#include <QQmlEngine>
+#include <QQmlContext>
+int main(int argc, char *argv[])
```
/n above

Reensef commented on plugins/interface/CMakeLists.txt:
```diff
@@ -5,13 +5,19 @@ project(
 )
 
-INCLUDE_SOURCES(${PROJECT_NAME})
+file(GLOB_RECURSE SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)
```
Почему macro был удалён?

bhaalgore committed:
resolving some mistakes

bhaalgore committed:
Merge branch 'master' into feature/add-logic-model

bhaalgore committed:
add some fixes

unitedtimur commented on plugins/logic/include/notify_model.h:
```diff
@@ -0,0 +1,80 @@
+        /*!
+         * \brief Метод, возвращающий количество уведомлений в модели
+         */
+        int count() const;
```
Добавь `\n` перед методами

unitedtimur commented on plugins/logic/include/notify_model.h:
```diff
@@ -0,0 +1,80 @@
+        int count() const;
+
+    private:
+        struct Notification
```
`Notification` не подойдёт. Должна быть абстрактная сущность, которая к примеру будет называться `AbstractNotification`, от которой каждый пользователь библиотеки будет наследоваться и дополнять или переопределять по необходимости методы их функционал. То есть надо сделать абстрактный класс с базовыми функциями по полям ,котоорые ты описал ниже, и уже унаследовавшись от него, сделать вашу имплементацию.

bhaalgore committed:
add abstract notification class

bhaalgore committed:
remove useless enum type

bhaalgore committed:
add implementation of notification class

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,16 @@
+class AbstractNotification
+{
+public:
+    AbstractNotification(QString title, QString description, int type);
```
`explicit`

bhaalgore committed:
naming fixes

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,29 @@
+         * \param description - строка, в которой храниться более подробное описание уведомления
+         * \param type - тип уведомления
+         */
+        AbstractNotification(QString title, QString description, int type);
```
`explicit`

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,29 @@
+     * \brief Класс, от которого могут наследоваться юзеры библиотеки, чтобы создать новый класс
+     * уведомления, если потребуется расширить функционал
+     */
+    class AbstractNotification
```
Ошибка, нет экспортирования символов. А также наследования от Qobject

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,29 @@
+        AbstractNotification(QString title, QString description, int type);
+        ~AbstractNotification() = default;
+
+        QString _title;
```
Приватные члены не должны быть видимыми!!!!!!

bhaalgore replied:
В новых комитах исправил это), добавил геттеры

unitedtimur commented on plugins/logic/include/notification_entity.h:
```diff
@@ -0,0 +1,17 @@
+    /*!
+     * \brief Имплементация класса, в который нужно вносить изменения, если это потребуется
+     */
+    class notification_entity final : public AbstractNotification
```
?

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,29 @@
+
+        QString _title;
+        QString _description;
+        int _type;
```
Тип уведомления должен быть вариативным. Желательно расширяемым `enum`

unitedtimur replied:
У меня есть такой класс.
![image](https://user-images.githubusercontent.com/30841985/232459290-8ae4b6ce-044b-4c7b-ba64-ad5584246c71.png)

Который потом используется так 
![image](https://user-images.githubusercontent.com/30841985/232459431-a8c1d04a-3301-4061-8492-b5791d49e803.png)

bhaalgore replied:
Я уже задумался об этом. Сейчас подумаю как это реализовать.

bhaalgore committed:
add getters

bhaalgore committed:
naming fixes

bhaalgore committed:
rework model class for interaction with notification class

bhaalgore committed:
add export to abstract notification

bhaalgore committed:
add new constructor and setters

bhaalgore committed:
fixes according with another argument in addNotification method

bhaalgore committed:
fixes according  with another type in QList template

bhaalgore committed:
add notifyType class

bhaalgore committed:
Merge branch 'master' into feature/add-logic-model

bhaalgore committed:
some fixes

bhaalgore committed:
delete useless constructor

bhaalgore committed:
fix

bhaalgore committed:
fix dedines naming

bhaalgore committed:
fix spacing

goshantiy committed:
gitignore

unitedtimur commented on plugins/logic/include/notification_entity.h:
```diff
@@ -0,0 +1,17 @@
+    class NotificationEntity final : public AbstractNotification
+    {
+    public:
+        NotificationEntity(QString title, QString description, int type);
```
`explicit`

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,52 @@
+        QString getTitle();
+        QString getDescription();
+        int getType();
+        //LogicPlugin::NotificationType getType();
```
?

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,52 @@
+         * \brief Геттеры для приватных полей класса
+         */
+        QString getTitle();
+        QString getDescription();
```
Как правило у нас стилистика `getters` пишится без `get`, дабы и так понятно если есть спецификатор `const`

bhaalgore committed:
fix getters naming due stylistics

bhaalgore committed:
rework of logic model

- Исправлены ошибки, указанные в conversations.
- Переработана логика взаимодействия класса AbstractNotification и NotificationType
- Исправлены методы NotificationType
- Добавлен метод, проверяющий существование уведомления по его id

goshantiy committed:
Merge remote-tracking branch 'origin/feature/add-logic-model' into feature/add-logic-model

goshantiy committed:
add logger

goshantiy committed:
little fix

goshantiy committed:
close db

goshantiy committed:
add logger to model

goshantiy committed:
naming

bhaalgore committed:
spacings fix

goshantiy committed:
Merge remote-tracking branch 'origin/feature/notification-logger' into feature/notification-logger

goshantiy committed:
brief

goshantiy committed:
fix brief

goshantiy committed:
change in query type to text

goshantiy committed:
brief

goshantiy committed:
one more brief

bhaalgore committed:
fix CI

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,53 @@
+         */
+        const QString Title();
+        const QString Description();
+        const QString Type();
```
`QString type() const;` и остально по аналогии, а также коммент у каждого должно быть.

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,53 @@
+        /*!
+         * \brief Сеттеры для приватных полей класса
+         */
+        void setTitle(QString title);
```
` void setTitle(const QString &title);` С комментариями аналогично.

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,53 @@
+        /*!
+         * \brief приватные поля класса, в них храняться название уведомление, краткое описание и id
+         * типа уведомления, к каждому id привязано конкретное название уведомления
+         */
```
На каждое поле свой комментарий, плюс они с большой буквы.

unitedtimur commented on plugins/logic/include/notification_type.h:
```diff
@@ -0,0 +1,49 @@
+         * храняться уведомления и соответсвующие им id
+         * \param name - название уведомления, которое требуется добавить
+         */
+        void addType(const QString &name);
```
Этот метод должен возвращать `std::optional<uint16_t>`

unitedtimur commented on plugins/logic/include/notification_logger.h:
```diff
@@ -0,0 +1,28 @@
+    };
+
+}
+#endif // NOTIFICATIONLOGGER_H
```
Ничего не описано, класс не читаемый.

unitedtimur commented on plugins/logic/include/notification_logger.h:
```diff
@@ -0,0 +1,28 @@
+#ifndef NOTIFICATIONLOGGER_H
+#define NOTIFICATIONLOGGER_H
```
Тоже не по канону.

unitedtimur commented on plugins/logic/include/notification_entity.h:
```diff
@@ -0,0 +1,17 @@
+    class NotificationEntity final : public AbstractNotification
+    {
+    public:
+       explicit NotificationEntity(QString title, QString description, int type);
```
См. какой тип в основе, это не `int`

unitedtimur commented on plugins/logic/src/abstract_notification.cpp:
```diff
@@ -0,0 +1,47 @@
+
+void LogicPlugin::AbstractNotification::setType(uint16_t typeID)
+{
+    if (LogicPlugin::NotificationType::isTypeExist(typeID))
```
!code style, должны быть операторные скобки.

unitedtimur commented on plugins/logic/include/notify_model.h:
```diff
@@ -0,0 +1,71 @@
+        /*!
+         * \brief Список, в котором хранится информация об всех уведомления в модели
+         */
+        QList<QPointer<LogicPlugin::AbstractNotification>> _notifications;
```
Весь код надо привести к код стайлу.

unitedtimur commented on plugins/logic/include/notify_model.h:
```diff
@@ -0,0 +1,71 @@
+     * \brief Класс, который определяет интерфейс, используемый для взаимодействия с компонентами
+     * gui
+     */
+    class LOGIC_PLUGIN_EXPORT NotificationModel : public QAbstractItemModel
```
`final`

unitedtimur commented on plugins/logic/src/notification_entity.cpp:
```diff
@@ -0,0 +1,4 @@
+#include "notification_entity.h"
+
+LogicPlugin::NotificationEntity::NotificationEntity(QString title, QString description, int type): AbstractNotification(title,  description, type)
```
!code style

unitedtimur commented on plugins/logic/src/notification_logger.cpp:
```diff
@@ -0,0 +1,37 @@
+    {
+        QSqlQuery query(_db);
+        return query.exec("CREATE TABLE IF NOT EXISTS notification(Type INT, Title TEXT, "
+                          "Description TEXT, Date DATETIME)");
```
См. как использовать `bind`

goshantiy replied:
bindValue в query нельзя использовать для column name
  QSqlQuery query(_db);
        query.prepare("CREATE TABLE IF NOT EXISTS notification(:type TEXT, :title TEXT, "
                      ":descript TEXT, :date DATETIME)");
        query.bindValue(":type", QVariant::fromValue(QString("Type")));
        query.bindValue(":title", QVariant::fromValue(QString("Title")));
        query.bindValue(":descript", QVariant::fromValue(QString("Description")));
        query.bindValue(":date", QVariant::fromValue(QString("Date")));
        return query.exec();
        Такой запрос не создаёт таблицу, обычный запрос строкой здесь работает. Может я что-то не понял?

unitedtimur replied:
Используй `prepare`

https://doc.qt.io/qt-6/qsqlquery.html#prepare

unitedtimur commented on plugins/logic/src/notification_type.cpp:
```diff
@@ -0,0 +1,33 @@
+
+bool LogicPlugin::NotificationType::isTypeExist(uint16_t value)
+{
+    for (auto it = _types.begin(); it != _types.end(); ++it) {
```
`std::find_if`

unitedtimur commented on plugins/logic/src/notification_type.cpp:
```diff
@@ -0,0 +1,33 @@
+
+const QString LogicPlugin::NotificationType::stringType(uint16_t value)
+{
+    for (auto it = _types.begin(); it != _types.end(); ++it) {
```
`std::find_if`

unitedtimur commented on plugins/logic/src/notification_type.cpp:
```diff
@@ -0,0 +1,33 @@
+
+void LogicPlugin::NotificationType::addType(const QString &name)
+{
+    if (name != "") {
```
```
if (name.isEmpty())
return;

const auto it = std::find_if ....
```

unitedtimur commented on plugins/logic/src/notify_model.cpp:
```diff
@@ -0,0 +1,92 @@
+QModelIndex
+LogicPlugin::NotificationModel::index(int row, int column, const QModelIndex &parent) const
+{
+    if (parent.isValid() || row < 0 || row >= _notifications.count() || column < 0 || column >= 3) {
```
Что за магические чиселки?

bhaalgore replied:
Переделали тип модели на QAbstractListModel, все проверки убрали

unitedtimur commented on plugins/logic/src/notify_model.cpp:
```diff
@@ -0,0 +1,92 @@
+int LogicPlugin::NotificationModel::columnCount(const QModelIndex &parent) const
+{
+    Q_UNUSED(parent);
+    return 3;
```
Аюсолютли неверно.

unitedtimur replied:
Или такая задумка?

unitedtimur commented on plugins/logic/src/notify_model.cpp:
```diff
@@ -0,0 +1,92 @@
+QVariant LogicPlugin::NotificationModel::data(const QModelIndex &index, int role) const
+{
+    if (!index.isValid() || index.row() < 0 || index.row() >= _notifications.count()
+        || index.column() < 0 || index.column() >= 3) {
```
Аналогично

goshantiy committed:
spacing

goshantiy committed:
naming

goshantiy committed:
brief

goshantiy committed:
QVariant in query fromValue

goshantiy committed:
update model

bhaalgore committed:
Merge remote-tracking branch 'origin/feature/notification-logger' into feature/add-logic-model

bhaalgore committed:
fix getters naming and add description

bhaalgore committed:
fix setters naming and add description

bhaalgore committed:
fix description of class members

bhaalgore committed:
stylistics fixes

unitedtimur commented on plugins/logic/src/notify_model.cpp:
```diff
@@ -0,0 +1,74 @@
+        case MessageRole:
+            return _notifications.at(index.row())->description();
+        case TypeRole:
+            return _notifications.at(index.row())->type();
```
Запрашивать `index.row()` always не хорошо, лучше вынести это в `const auto` переменную выше.

unitedtimur commented on plugins/logic/src/notify_model.cpp:
```diff
@@ -0,0 +1,74 @@
+            return QVariant();
+        }
+    }
+    return QVariant();
```
Здесь `return` не имеет смысла, ибо в `default` у тебя уже есть выход.

unitedtimur commented on plugins/logic/src/notify_model.cpp:
```diff
@@ -0,0 +1,74 @@
+
+QVariant LogicPlugin::NotificationModel::data(const QModelIndex &index, int role) const
+{
+    if (index.row() <= _notifications.count()) {
```
```
if (index.row > _notifications.size())
return;
```

unitedtimur commented on plugins/logic/src/notify_model.cpp:
```diff
@@ -0,0 +1,74 @@
+    logger.insert(notification->type(), notification->title(), notification->description(),
+                  QDate::currentDate());
+    endInsertRows();
+    emit dataChanged(
```
Избыточно вызывать сигнал об изменении, если до этого уже используется `beginInsertRows` и тд

goshantiy committed:
minor changes

goshantiy committed:
Merge branch 'feature/add-logic-model' into feature/notification-logger

bhaalgore committed:
change qmap to std::map and rewrite methods

bhaalgore committed:
Merge remote-tracking branch 'origin/feature/notification-logger' into feature/add-logic-model

bhaalgore committed:
fix naming

bhaalgore committed:
fix arguments types

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,77 @@
+         * \brief Приватное поле класса, хранящее id типа уведомления, к каждому id привязано
+         * конкретное название уведомления
+         */
+        uint16_t _typeID;
```
Зачем одна переменная на весь плагин?

unitedtimur commented on plugins/logic/include/abstract_notification.h:
```diff
@@ -0,0 +1,77 @@
+        /*!
+         * \brief Приватное поле класса, хранящее краткое описание уведомления
+         */
+        QString _description;
```
Аналогично.

unitedtimur commented on plugins/logic/include/notification_logger.h:
```diff
@@ -0,0 +1,71 @@
+        ~NotificationLogger() = default;
+        NotificationLogger(const NotificationLogger &) = delete;
+        NotificationLogger &operator=(NotificationLogger) = delete;
+        /*!
```
add `\n` above

bhaalgore committed:
change addType method

unitedtimur commented on plugins/logic/include/notification_type.h:
```diff
@@ -0,0 +1,49 @@
+        /*!
+         * \brief _nTypes - количество уведомлений в структуре (по умолчанию их 4)
+         */
+        uint16_t _nTypes = 4;
```
`uint16_t _typeCount { 4 }`

unitedtimur commented on plugins/logic/include/notification_type.h:
```diff
@@ -0,0 +1,49 @@
+         * \brief _types - структура, в которой хранятся пары значений (название уведомления и его
+         * id)
+         */
+        static std::map<QString, uint16_t> _types;
```
Если есть такая структура, то поле ниже не нужно. + переменная не должна быть статической.

unitedtimur commented on plugins/logic/src/notification_type.cpp:
```diff
@@ -0,0 +1,41 @@
+                                 });
+    if (it != _types.end()) {
+        return it->first;
+    } else {
```
`else` не нужно.

unitedtimur commented on plugins/logic/src/notification_type.cpp:
```diff
@@ -0,0 +1,41 @@
+const QString LogicPlugin::NotificationType::stringType(const uint16_t &value)
+{
+    const auto it = std::find_if(_types.begin(), _types.end(),
+                                 [&value](const std::pair<QString, uint16_t> &element) {
```
`const auto &type`

unitedtimur commented on plugins/logic/src/notification_type.cpp:
```diff
@@ -0,0 +1,41 @@
+                                 });
+    if (it != _types.end()) {
+        return true;
+    } else {
```
Аналогично.

goshantiy committed:
\n

goshantiy committed:
const auto row

goshantiy committed:
naming

bhaalgore committed:
fix

bhaalgore committed:
rename class

goshantiy committed:
include optional

bhaalgore committed:
add file extansion type to db name

bhaalgore committed:
Merge remote-tracking branch 'origin/feature/add-logic-model' into feature/add-logic-model

bhaalgore committed:
rem useless class member

bhaalgore committed:
small fix
